src/tools/kata-ctl/src/exec.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{
+    io::{self, BufRead, BufReader, Read, Write},
+    os::unix::{
+        io::{AsRawFd, RawFd},
+        net::UnixStream,
+    },
+    path::PathBuf,
+    time::Duration,
+};
+
+use anyhow::anyhow;
+use glob::glob;
+use serde::{Deserialize, Serialize};
+use slog::debug;
+use vmm_sys_util::terminal::Terminal;
+
+const KATA_RUN_PATH: &str = "/run/kata";
+const KATA_HYBRID_VSOCK: &str = "kata.hvsock";
+const CONNECT_CMD: &str = "CONNECT";
+const OK_CMD: &str = "OK";
+
+// 5 seconds
+const KATA_AGENT_VSOCK_TIMEOUT: u64 = 5;
+
+// Convenience macro to obtain the scope logger
+#[macro_export]
+macro_rules! sl {
+    () => {
+        slog_scope::logger()
+    };
+}
+
+#[derive(Debug)]
+pub enum Error {
+    EpollWait(io::Error),
+    EpollCreate(io::Error),
+    EpollAdd(io::Error),
+    SocketWrite(io::Error),
+    StdioErr(io::Error),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Config {
+    pub server_address: String,
+    pub hybrid_vsock: bool,
+    pub hybrid_vsock_port: u64,
+    pub bundle_dir: String,
+    pub interactive: bool,
+}
+
+enum EpollDispatch {
+    Stdin,
+    ServerSock,
+}
+
+struct EpollContext {
+    epoll_raw_fd: RawFd,
+    stdin_index: u64,
+    dispatch_table: Vec<EpollDispatch>,
+    stdin_handle: io::Stdin,
+    debug_console_sock: Option<UnixStream>,
+}
+
+impl EpollContext {
+    fn new() -> Result<Self> {
+        let epoll_raw_fd = epoll::create(true).map_err(Error::EpollCreate)?;
+        let dispatch_table = Vec::with_capacity(2);
```

Comment:
Seems there is no need to initialize dispatch_table with capacity 2. Elements added to dispatch table are all using `.push()` in the later operation

Reply:
> > This is my configuration-dragonball.toml's kernel params:
> > kernel_params = "console=ttyS1 agent.log_vport=1025 agent.debug_console agent.debug_console_vport=1026"
> 
> @Apokleos I think `agent.debug_console`and `agent.debug_console.vport=1026` should be added by users in `kernel_params` is not a good idea. How about appending these two params if `enable_debug` is true in agent's configuration? (But by this way, the user will not allowed to choose a vport then. So I wonder how is it implemented in `kata-runtime `and `runtime-go`?)

Hi @YushuoEdge 
1). `agent.debug_console`and `agent.debug_console.vport=1026` should be added by users in `kernel_params`. And agent will do debug console handler with debug_console[y|n] and debug_console_vport. If not, default vport is 0.
2). `enable_debug` is for debug message not for entering guest with debug_console.